services/accountConfirmation.service.js

```diff
@@ -0,0 +1,117 @@
+"use strict";
+const logger = require("./logger.service");
+const AccountConfirmation = require("../models/accountConfirmationToken.model");
+const Constants = require("../constants");
+const jwt = require("jsonwebtoken");
+
+/**
+ * @function findByAccountId
+ * @param {ObjectId} accountId
+ * @return {DocumentQuery} The document query will resolve to either account confirmation or null.
+ * @description Finds an account confirmation document by accountId.
+ */
+function findByAccountId (accountId) {
+    const TAG = `[ AccountConfirmation Reset Service # findByAccountId ]`;
+    return AccountConfirmation.findOne({
+        accountId: accountId
+    }, logger.queryCallbackFactory(TAG, "AccountConfirmation", "accountId"));
+}
+
+/**
+ * @function findById
+ * @param {ObjectId} id
+ * @return {DocumentQuery} The document query will resolve to either account confirmation or null.
+ * @description Finds an account by mongoID.
+ */
+function findById (id) {
+    const TAG = `[ AccountConfirmation Service # findById ]`;
+    return AccountConfirmation.findById(id, logger.queryCallbackFactory(TAG, "AccountConfirmation", "mongoId"));
+}
+
+/**
+ * Creates Account Confirmation document in the database
+ * @param {String} type the type of user which to create the token for
+ * @param {String} email
+ * @param {ObjectId} accountId optional accountId parameter to link to account, optional when token is being made for not a hacker
+ * @returns {Promise.<*>}
+ */
+async function create (type, email, accountId) {
+    //Create new instance of account confirmation
+    const newAccountToken = AccountConfirmation({
+        accountType: type,
+        email: email
+    });
+    if(accountId !== undefined){
+        newAccountToken.accountId = accountId;
+    }
+    return newAccountToken.save();
+}
+
+/**
+ * Generates JWT for Confirming account
+ * @param {ObjectId} accountConfirmationId
+ * @param {ObjectId} accountId
+ * @returns {string} JWT Token containing accountId and accountConfirmationId
+ */
+function generateToken (accountConfirmationId, accountId) {
+    const token = jwt.sign({
+        accountConfirmationId: accountConfirmationId,
+        accountId: accountId
+    }, process.env.JWT_CONFIRM_ACC_SECRET, {
+        expiresIn: "7 day"
+    });
+    return token;
+}
+
+/**
+ * Generates the link that the user will use to access the page to finish account creation
+ * @param {'http'|'https'} httpOrHttps 
+ * @param {string} domain the domain of the current
+ * @param {string} type the model that the 
+ * @param {string} token the reset token
+ * @returns {string} the string, of form: [http|https]://{domain}/{model}/reset?token={token}
+ */
+function generateTokenLink (httpOrHttps, domain, type, token) {
+    const link = `${httpOrHttps}://${domain}/${type}/create?token=${token}`;
```

Comment:
We might have to change this so that it's not `/create`, but that's a future thing

Reply:
Noted